src/oop_ext/_type_checker_fixture.py

```diff
@@ -0,0 +1,90 @@
+# mypy: disallow-untyped-defs
+# mypy: disallow-any-decorated
+import os
+import re
+from textwrap import dedent
+
+import mypy.api
+from pathlib import Path
+from typing import List, Tuple
+
+import attr
+import pytest
+
+
+@attr.s(auto_attribs=True)
+class _Result:
+    """
+    Encapsulates the result of a call to ``mypy.api``, providing helpful functions to check
+    that output.
+    """
+
+    output: Tuple[str, str, int]
+
+    def assert_errors(self, messages: List[str]) -> None:
+        assert self.error_report == ""
+        lines = self.report_lines
+        assert len(lines) == len(
+            messages
+        ), f"Expected {len(messages)} failures, got {len(lines)}:\n" + "\n".join(lines)
+        for index, (obtained, expected) in enumerate(zip(lines, messages)):
+            m = re.search(expected, obtained)
+            assert m is not None, (
+                f"Expected regex at index {index}:\n"
+                f"  {expected}\n"
+                f"did not match:\n"
+                f"  {obtained}\n"
+                f"(note: use re.escape() to escape regex special characters)"
+            )
+
+    def assert_ok(self) -> None:
+        assert len(self.report_lines) == 0, "Expected no errors, got:\n " + "\n".join(
+            self.report_lines
+        )
+        assert self.exit_status == 0
+
+    @property
+    def normal_report(self) -> str:
+        return self.output[0]
+
+    @property
+    def error_report(self) -> str:
+        return self.output[1]
+
+    @property
+    def exit_status(self) -> int:
+        return self.output[2]
+
+    @property
+    def report_lines(self) -> List[str]:
+        lines = [x.strip() for x in self.normal_report.split("\n") if x.strip()]
+        # Drop last line (summary).
+        return lines[:-1]
+
+
+@attr.s(auto_attribs=True)
+class TypeCheckerFixture:
```

Comment:
Hey @bluetech what do you think of this? Overkill, or worthwhile turning into a plugin?

Reply:
I think a pytest plugin is definitely useful for projects which want to verify their typing in a precise way.

I am aware of existing plugin https://github.com/davidfritzsche/pytest-mypy-testing, haven't checked its implementation.

Another comparison is to mypy's own tests, which use files like this: https://github.com/python/mypy/blob/master/test-data/unit/check-unreachable-code.test. They use pytest but I haven't checked how they run these.

BTW (unrelated), in the future you might be able to replace the 5 Callback types one just one using [Variadic Generics](https://www.python.org/dev/peps/pep-0646/), if it get accepted and implemented.

Reply:
> I am aware of existing plugin davidfritzsche/pytest-mypy-testing, haven't checked its implementation.

> Another comparison is to mypy's own tests, which use files like this: https://github.com/python/mypy/blob/master/test-data/unit/check-unreachable-code.test. They use pytest but I haven't checked how they run these.

Cool, I did a cursory search and missed those, I will check them out, thanks!

> BTW (unrelated), in the future you might be able to replace the 5 Callback types one just one using Variadic Generics, if it get accepted and implemented.

Yes, I'm aware of that, let's hope that gets in! 🤞 

Thanks for the feedback. 👍